### PR TITLE
Add simple fast inventory

### PR DESCRIPTION
**Pull request by rubenwardy**
Allows mods to define tabs for the 'i' inventory, even when creative mode is off.

``` lua
sfinv.register_page("mymod:crafting2", {
    title = "Crafting2",
    get = function(self, player, context, vars)
        return sfinv.make_formspec(player, context, [[
                list[current_player;craft;1.75,0.5;3,3;]
                list[current_player;craftpreview;5.75,1.5;1,1;]
                image[4.75,1.5;1,1;gui_furnace_arrow_bg.png^[transformR270]
                listring[current_player;main]
                listring[current_player;craft]
                image[0,4.75;1,1;gui_hb_bg.png]
                image[1,4.75;1,1;gui_hb_bg.png]
                image[2,4.75;1,1;gui_hb_bg.png]
                image[3,4.75;1,1;gui_hb_bg.png]
                image[4,4.75;1,1;gui_hb_bg.png]
                image[5,4.75;1,1;gui_hb_bg.png]
                image[6,4.75;1,1;gui_hb_bg.png]
                image[7,4.75;1,1;gui_hb_bg.png]
            ]], true)
    end,
    on_player_receive_fields = function(self, player, context, fields)
        print(dump(fields))
    end
})
```
## Potential Problems
- Allowing mods to inject formspec into existing pages - eg: armor mods. This could be done currently, and would work with the default layout given, but if a layout used containers then the injected content would not be shifted down. TL;DR: it is possible for mods to inject content, but it would break layout support for that page a little
## Planned improvements

When this PR is merged, I'll work on:
- Unified inventory support using a wrapper
- Creative inventory improvements:
  - Sort by description (more natural)
  - Add sub sections for nodes, such as natural items / decorations / mechanics / etc
  - ...

### CODE REVIEW

File: mods/sfinv/init.lua [hunk 1]
```diff
@@ -0,0 +1,25 @@
+dofile(minetest.get_modpath("sfinv") .. "/api.lua")
+
+sfinv.register_page("sfinv:crafting", {
+	title = "Crafting",
+	is_in_nav = function(self, player, context)
```
Comment (rubenwardy) [inline mods/sfinv/init.lua]: This should be removed as not needed

File: game_api.txt [hunk 2]
```diff
@@ -338,6 +348,104 @@ To use it, add the `on_screwdriver` function to the node definition.
...
+def is a table containing:
+
+* `title` - human readable page name (required)
+* `get(self, player, context, vars)` - returns a formspec string. See formspec variables. (required)
+* `is_in_nav(self, player, context)` - return true if it appears in tab header
+* `on_player_receive_fields(self, player, context, fields)` - on formspec submit.
+* `on_enter(self, player, context)` - called when the player changes pages, usually using the tabs.
+* `on_leave(self, player, context)` - when leaving this page to go to another, called before other's on_enter
+
+### Makeing formspecs using variables
```
Comment (rubenwardy) [inline game_api.txt]: typo

**Comment by sofar** [comment]
I've been waiting for something like this, so, I'm very positive about a change like this.

**Comment by benrob0329** [comment]
+1

File: mods/creative/inventory.lua [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+			local player_name = player:get_player_name()
+			creative.update_creative_inventory(player_name, items)
+			local inv = player_inventory[player_name]
+			local start_i = inv.start_i or 0
+			local pagenum = math.floor(start_i / (3*8) + 1)
+			local pagemax = math.ceil(inv.size / (3*8))
+			return [[
+					size[8,8.6]
+					bgcolor[#080808BB;true]
+					background[5,5;1,1;gui_formbg.png;true]
```
Comment (t4im) [inline mods/creative/inventory.lua]: Aren't these overwritten later with `default.gui_bg .. default.gui_bg_img ..`?

Comment (rubenwardy) [inline mods/creative/inventory.lua]: Background is an image which is drawn below all over elements, this background image is at (5, 5) and has size (1, 1). I imagine that default.gui_bg covers the entire window

File: mods/creative/inventory.lua [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+				inv.start_i = start_i
+				sfinv.set(player, context)
+			end
+		end
+	})
+end
+
+minetest.register_on_joinplayer(function(player)
+	-- If in creative mode, modify player's inventory forms
+	if not creative_mode then
```
Comment (rubenwardy) [inline mods/creative/inventory.lua]: bad code style (copy pasta from old mod)

File: mods/sfinv/api.lua [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+		local tid = tonumber(fields.tabs)
+		if tid and tid > 0 then
+			local id = context.nav[tid]
+			local page = sfinv.pages[id]
+			if id and page then
+				print(name .. " views sfinv/" .. id)
+
+				local oldpage = sfinv.pages[context.page]
+				if oldpage and oldpage.on_leave then
+					oldpage:on_leave(player, content)
```
Comment (t4im) [inline mods/sfinv/api.lua]: Content is `nil` here and further down, is that supposed to be context?

File: mods/creative/inventory.lua [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+				]] ..
+				vars.nav ..
+				[[
+					listcolors[#00000069;#5A5A5A;#141318;#30434C;#FFF]
+					list[current_player;main;0,4.7;8,1;]
+					list[current_player;main;0,5.85;8,3;8]
+					list[detached:creative_trash;main;4,3.3;1,1;]
+					listring[]
+					tablecolumns[color;text;color;text]
+					tableoptions[background=#00000000;highlight=#00000000;border=false]
```
Comment (t4im) [inline mods/creative/inventory.lua]: If the tablecolumns, options and table are used for coloring the page, wouldn't a label with `minetest.colorize` make things a little more simple?

Comment (rubenwardy) [inline mods/creative/inventory.lua]: Copy pasta from old creative mod, might as well fix here though

File: mods/sfinv/api.lua (hunk 5 above)
Comment (rubenwardy) [inline mods/sfinv/api.lua]: good catch, it is

**Comment by rubenwardy** [comment]
Fixed all issues except page number coloring

File: game_api.txt [hunk 7]
```diff
@@ -338,6 +348,104 @@ To use it, add the `on_screwdriver` function to the node definition.
...
+
+### sfinv.register_page
+
+sfinv.register_page(name, def)
+
+def is a table containing:
+
+* `title` - human readable page name (required)
+* `get(self, player, context, vars)` - returns a formspec string. See formspec variables. (required)
+* `is_in_nav(self, player, context)` - return true if it appears in tab header
```
Comment (SmallJoker) [inline game_api.txt]: This looks a bit confusing, how about `Return true to show it in the tab header`?

**Comment by Ekdohibs** [comment]
I haven't tested it, but I'm very positive about the idea as well.
While you're at it, could you please do the page number coloring change, as t4im suggested?

File: game_api.txt (hunk 7 above)
Comment (rubenwardy) [inline game_api.txt]: fixed (hopefully to your satisfaction)

**Comment by rubenwardy** [comment]
It now uses a label and colorize instead of a table.
However the text is not center aligned.
Does anyone know how to make it so?
If not, then it will have to be reverted

**Comment by C1ffisme** [comment]
Very useful for things like armor, etc.

We can finally get rid of inventory_plus and it's like!

**Comment by Ekdohibs** [comment]
@rubenwardy Hm, wouldn't there be problems with the text overlapping other buttons if it's center-aligned?

**Comment by rubenwardy** [comment]
Ideally you'd be able to specify a rectangle for the text to be in, then center align it there

**Comment by Ekdohibs** [comment]
Hm, I don't know if this is possible though. I think keeping it left-aligned is good anyway.

**Comment by rubenwardy** [comment]
Is using a table that much of a problem? It's a bit overkill, but at least it looks right

**Comment by rubenwardy** [comment]
Please note than an older version of this code has been running on the capture the flag server with no noticed performance issues

**Comment by t4im** [comment]
> with no noticed performance issues.

If it's performance you want to know about, let me give it a spin in the current profiler version

```md
 instrumentation                                         |    min µs |    max µs |    avg µs | min % | max % | avg %
---------------------------------------------------------+-----------+-----------+-----------+-------+-------+-------
 sfinv:                                                  |         0 |      7946 |        68 |   0.0 |  99.4 |   2.2
  - on_player_receive_fields[1] ........................ |         0 |      7946 |        71 |   0.0 |  99.4 |   2.3
  - on_joinplayer[1] ................................... |         0 |        10 |         0 |   0.0 |   0.2 |   0.0
 creative:                                               |         0 |      5492 |         1 |   0.0 |  90.2 |   0.0
  - on_joinplayer[1] ................................... |         0 |      5492 |         1 |   0.0 |  90.2 |   0.0
```

I'd say that's quite alright. But of course the numbers would be more interesting when viewed over a day or so on an active server.

**Comment by paramat** [comment]
I have no opinion on this so am neutral. I can't review as am useless with inventory / formspec stuff.
Luckily it seems you have the attention of 2 mtgame devs already.

**Comment by 0-afflatus** [comment]
Great idea, don't have time to review properly.

**Comment by Fixer-007** [comment]
This should be a nice thing, I'm tired of clicking via buttons in inventory plus, should be much faster and easier with tabs :+1:

**Comment by rubenwardy** [comment]
rebased

**Comment by rubenwardy** [comment]
Found something weird: searching for "bone" gets no results even though bone is seen on the first page, and has bone in the description. Very weird.

**Comment by rubenwardy** [comment]
Ok, bones appear if you click left after searching for it. This makes me think that the start idx isn't being reset correctly.

File: mods/sfinv/api.lua [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+			}
+			sfinv.contexts[name] = context
+		end
+	end
+
+	local fs = sfinv.get_formspec(player, context)
+	player:set_inventory_formspec(fs)
+end
+
+minetest.register_on_joinplayer(function(player)
```
Comment (t4im) [inline mods/sfinv/api.lua]: Wouldn't it make sense to free the contexts on leave player, to free unused memory?

Comment (rubenwardy) [inline mods/sfinv/api.lua]: done

**Comment by paramat** [comment]
Since you're a dev this can be merged once you get 1 approval from someone else.

File: game_api.txt [hunk 9]
```diff
@@ -375,6 +385,96 @@ set a players home position and teleport a player to home position.
...
+		listring[current_player;craft]
+		image[0,4.25;1,1;gui_hb_bg.png]
+		image[1,4.25;1,1;gui_hb_bg.png]
+		image[2,4.25;1,1;gui_hb_bg.png]
+		image[3,4.25;1,1;gui_hb_bg.png]
+		image[4,4.25;1,1;gui_hb_bg.png]
+		image[5,4.25;1,1;gui_hb_bg.png]
+		image[6,4.25;1,1;gui_hb_bg.png]
+		image[7,4.25;1,1;gui_hb_bg.png]
+	]], true)
```
Comment (rubenwardy) [inline game_api.txt]: needs to fully document make_formspec

File: game_api.txt [hunk 10]
```diff
@@ -375,6 +385,96 @@ set a players home position and teleport a player to home position.
...
+        local tmp = {
+            theme_main,
+            sfinv.get_nav_fs(player, context, context.nav_titles, context.nav_idx),
+            content
+        }
+        if show_inv then
+            tmp[4] = theme_inv
+        end
+        return table.concat(tmp, "")
+    end
```
Comment (rubenwardy) [inline game_api.txt]: inconsistent indentation

File: mods/sfinv/README.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
+# Simple Fast Inventory
+
+![SFINV Screeny](https://cdn.pbrd.co/images/1yQhd1TI.png)
+
+A cleaner, simpler, solution to having an advanced inventory in Minetest.  
+Formspec style based on the creative inventory.
```
Comment (rubenwardy) [inline mods/sfinv/README.md]: No longer relevant

File: mods/sfinv/api.lua [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+	-- Handle Events
+	if fields.tabs and context.nav then
+		local tid = tonumber(fields.tabs)
+		if tid and tid > 0 then
+			local id = context.nav[tid]
+			local page = sfinv.pages[id]
+			if id and page then
+				print(name .. " views sfinv/" .. id)
+
+				-- TODO: on_leave
```
Comment (rubenwardy) [inline mods/sfinv/api.lua]: remove

File: mods/sfinv/api.lua [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+function sfinv.override_page(name, def)
+	if not name or not def then
+		error("Invalid sfinv page override. Requires a name & def")
+	end
+	local page = sfinv.pages[name]
+	if not page then
+		error("sfinv: attempt to override page " .. dump(name) .. " which does not exist.")
+	end
+	for key, value in pairs(t) do
```
Comment (rubenwardy) [inline mods/sfinv/api.lua]: def, not t

File: mods/creative/inventory.lua [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+	creative.update_creative_inventory(player_name, minetest.registered_items)
+end
+
+function creative.update_creative_inventory(player_name, tab_content)
+	local creative_list = {}
+	local player_inv = minetest.get_inventory({type = "detached", name = "creative_" .. player_name})
+	local inv = player_inventory[player_name]
+
+	for name, def in pairs(tab_content) do
+		if not (def.groups.not_in_creative_inventory == 1) and
```
Comment (t4im) [inline mods/creative/inventory.lua]: You could use this overhaul to change the misplaced `not_in_creative_inventory` group into a node definition property.
Neither will there be a reason to use this group in a crafting recipe, nor to find nodes of this type on the map, nor to dig them differently and all the other things group can be used to group types of nodes instead of individual node properties

**Comment by sofar** [comment]
@stujones11 I'd assume you want to review this? This would obsolete/replace the need for unified_inventory I think.

File: mods/creative/inventory.lua (hunk 14 above)
Comment (rubenwardy) [inline mods/creative/inventory.lua]: this certainly needs to be discussed. I'm not entirely sure what groups are meant to be

File: game_api.txt [hunk 15]
```diff
@@ -375,6 +385,98 @@ set a players home position and teleport a player to home position.
...
+
+Simply override this function to change the navigation:
+
+	function sfinv.get_nav_fs(player, context, nav, current_idx)
+		return "navformspec"
+	end
+
+And override this function to change the layout:
+
+	function sfinv.make_formspec(player, context, content, show_inv)
```
Comment (rubenwardy) [inline game_api.txt]: + `, size)`

File: mods/creative/inventory.lua (hunk 14 above)
Comment (rubenwardy) [inline mods/creative/inventory.lua]: I feel like this is an unrelated change that would be better in another PR

**Comment by rubenwardy** [comment]
Self approved :+1:
I've tested this by writing mods that use it, and scouring over the code many times to check

File: mods/sfinv/api.lua [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+	end
+
+	-- Handle Events
+	if fields.tabs and context.nav then
+		local tid = tonumber(fields.tabs)
+		if tid and tid > 0 then
+			local id = context.nav[tid]
+			local page = sfinv.pages[id]
+			if id and page then
+				print(name .. " views sfinv/" .. id)
```
Comment (t4im) [inline mods/sfinv/api.lua]: Forgotten debug statement.

File: mods/sfinv/api.lua [hunk 17]
```diff
@@ -0,0 +1,167 @@
+sfinv = {
+	pages = {},
+	pages_unordered = {},
+	homepage_name = "sfinv:crafting",
+	contexts = {},
+	enabled = true
+}
+
+function sfinv.register_page(name, def)
+	if not name or not def or not def.get then
```
Comment (sofar) [inline mods/sfinv/api.lua]: Since `error()` is a fatal message, with a full server crash, I prefer `assert()` instead, since it's shorter and produces the same effect (halt) and backtrace.

So here I'd write:

```
assert(name)
assert(def)
assert(def.get)
assert(not sfinv.pages[name])
```
benefit of this is that if def.get is `nil` you'd see the assert for exactly that, not just an error for three things combined.

Comment (t4im) [inline mods/sfinv/api.lua]: On the other hand, `error()` allows the stack level to be defined as additional parameter, making it clear, where the error really happens.

```lua
error("Invalid sfinv page. Requires a name & def, and a get function in def", 2)
```

Comment (sofar) [inline mods/sfinv/api.lua]: But he'd still have to write 12 lines of code to distinguish between any of the possible failed input values. And a mod developer should be able to understand and debug failed asserts :)

There's nothing wrong with `error()` as a function, but it shouldn't replace `assert()` where asserts could do the job IMHO.

Comment (rubenwardy) [inline mods/sfinv/api.lua]: much better, didn't realise this existed. Thanks!

File: mods/sfinv/api.lua [hunk 18]
```diff
@@ -7,12 +7,11 @@ sfinv = {
...
-		error("Invalid sfinv page. Requires a name & def, and a get function in def")
-	end
-	if sfinv.pages[name] then
-		error("sfinv: attempt to register already existing page " .. dump(name))
-	end
+	assert(name, "Invalid sfinv page. Requires a name")
+	assert(def, "Invalid sfinv page. Requires a def[inition] table")
+	assert(def.get, "Invalid sfinv page. Def requires a get function.")
+	assert(not sfinv.pages[name], "Attempt to register already existing sfinv page " .. dump(name))
+
```
Comment (sofar) [inline mods/sfinv/api.lua]: yeah, that works!

**Comment by paramat** [comment]
@sofar @t4im perhaps this can be merged now?

File: mods/creative/inventory.lua [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+			if fields.creative_clear then
+				inv.start_i = 0
+				inv.filter = ""
+				creative.update_creative_inventory(player_name, items)
+				sfinv.set(player, context)
+			elseif fields.creative_search or
+					fields.key_enter_field == "creative_filter" then
+				inv.start_i = 0
+				inv.filter = fields.creative_filter:lower()
```
Comment (t4im) [inline mods/creative/inventory.lua]: `fields.creative_filter` is `nil`, at least for older clients, that don't have your formspec changes yet.
The dependency on a recent client might be acceptable, but it shouldn't crash a server, as it does here.

File: mods/sfinv/api.lua [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+		end
+	end
+
+	local fs = sfinv.get_formspec(player, context)
+	player:set_inventory_formspec(fs)
+end
+
+minetest.register_on_joinplayer(function(player)
+	if sfinv.enabled then
+		minetest.after(0.5, function()
```
Comment (t4im) [inline mods/sfinv/api.lua]: What is the `minetest.after` for?

The delayed initialization leads to the creative inventory to override any custom inventories on every join, including unified inventory.

If that is intentional, you should probably mention in the PR text, how inventory mods have to be updated to work.

MTG will essentially remove `set_inventory_formspec` from the API by rendering it resultless, which i think is noteworthy in the documentation as well, and should be accompanied by replacing it with a function that throws a warning when used by other mods, so that issues caused by this can be resolved swiftly.

Alternatively you could replace it with a function that will deactivate sfinv when being used. (sfinv itself can always keep the original function for its own use and a reset) That way you could even remain backward compatible.

File: mods/creative/inventory.lua (hunk 19 above)
Comment (t4im) [inline mods/creative/inventory.lua]: To produce the crash: click the search field on a client before the formspec changes.

**Comment by rubenwardy** [comment]
There's a pr open would allows ui to disable sfinv. It's as simple as

```lua
if minetest.global_exists("sfinv") then
    sfinv.enabled = false
end
```

**Comment by paramat** [comment]
Sorry to remove the 'one approval' label, the way we currently work is that a core dev author's approval is not labelled (personally i think there's good reason to label it but it upsets several devs). The author's self-approval still of course counts towards the 2 needed.

File: mods/creative/inventory.lua (hunk 19 above)
Comment (rubenwardy) [inline mods/creative/inventory.lua]: I'm not sure how to reproduce: do I need an old version of MT? If so, at which commit? What do you mean by:

> before the formspec changes

**Comment by rubenwardy** [comment]
> the way we currently work is that a core dev author's approval is not labelled (personally i think there's good reason to label it but it upsets several devs)

That's strange. The label system imo is handy for telling how many more approvals a PR needs to be merged. I guess instead it says how many have reviewed

File: mods/creative/inventory.lua (hunk 19 above)
Comment (t4im) [inline mods/creative/inventory.lua]: I assume, that it is related to the formspec changes of 4330c63 and e10fee0. Personally i experienced it with d767f02, which is shortly before introduction of those.

Comment (rubenwardy) [inline mods/creative/inventory.lua]: Ah! This is due to the field disappearing. See #4566

Comment (rubenwardy) [inline mods/creative/inventory.lua]: This issue is also present in current MTG

**Comment by paramat** [comment]
Indeed, but i gave up arguing the point.

**Comment by sofar** [comment]
@rubenwardy we should get this back on track again.

File: mods/sfinv/api.lua (hunk 20 above)
Comment (rubenwardy) [inline mods/sfinv/api.lua]: It is not possible to override set_inventory_formspec

**Comment by tobyplowy** [comment]
+1

**Comment by paramat** [comment]
2 weeks to freeze, want this in release?

**Comment by sofar** [comment]
I'd really like to get this in, I think it'll be a good step forward for mods.

**Comment by rubenwardy** [comment]
Does anyone have any more objections/points?

It isn't possible to modify set_inventory_formspec, the mods would have to do sfinv.enabled = false

**Comment by paramat** [comment]
Sorry i can't comment being useless with inventory stuff.
Conflict with creative?

**Comment by tobyplowy** [comment]
@rubenwardy @sofar @paramat Will it get there for 0.4.15?

**Comment by sofar** [comment]
Works ok. Did the trash icon background disappear in this PR?

**Comment by sofar** [comment]
I'd recommend a huge squash+merge. 37 commits is way too much. As squashed PR it's not that big
```
 game_api.txt                |  106 +++++++++++++++++++
 mods/creative/depends.txt   |    1 
 mods/creative/init.lua      |  237 --------------------------------------------
 mods/creative/inventory.lua |  170 +++++++++++++++++++++++++++++++
 mods/sfinv/README.md        |   21 +++
 mods/sfinv/api.lua          |  158 +++++++++++++++++++++++++++++
 mods/sfinv/depends.txt      |    1 
 mods/sfinv/init.lua         |   22 ++++
 8 files changed, 482 insertions(+), 234 deletions(-)
```

**Comment by rubenwardy** [comment]
Yes, I am going to squash - I was leaving it until later

**Comment by rubenwardy** [comment]
done

**Comment by rubenwardy** [comment]
readded trash background

**Comment by sofar** [comment]
:+1:

**Comment by paramat** [comment]
d42ae715b8a4e5a46a590b0f97683855e5985cc1

**Comment by paramat** [comment]
But,
```
-if creative_mode then
-	local digtime = 42
-	local caps = {times = {digtime, digtime, digtime}, uses = 0, maxlevel = 256}
+if minetest.setting_getbool("creative_mode") then
+	local digtime = 0.5
+	local caps = {times = {digtime, digtime, digtime}, uses = 0, maxlevel = 3}
```
Why was this changed? 42 and 256 are intentional @rubenwardy

**Comment by rubenwardy** [comment]
Sfinv was started before that change, so that was a rebasing error

**Comment by rubenwardy** [comment]
The problems with email are due to that mod using an outdated sfinv api. Will fix later

**Comment by paramat** [comment]
Ok.
